docs/configuration.rst
Configuration file syntax
=========================

The default and preferred file name is ``.launchpad.yaml`` (with the leading
dot to keep it discreetly out of the way).

The following documentation makes reference to various behaviours of
Launchpad.  At the time of writing these have been designed but not yet
implemented.

.. _identifiers:

Identifiers
-----------

Several configuration file keys accept *identifiers*: these must begin with
an ASCII lower-case letter or number, must be at least two characters long,
and may only contain ASCII lower-case letters or numbers, ``+``, ``.``,
``_``, or ``-``.

Top-level configuration
-----------------------

``pipeline`` (required)
     List of stages; each stage is either a job name or a list of job names.
     Stages are executed in series, and subsequent stages only execute if
     previous stages succeeded.
     (We plan to allow parallel job execution within each stage in the future,
     but that hasn't yet been implemented.)

     Job names are :ref:`identifiers <identifiers>`.

``jobs`` (required)
     Mapping of job names (:ref:`identifiers <identifiers>`) to job
     definitions.

``license`` (optional)
     The :ref:`license <license-properties>` info for the given repository can
     be configured either via an
     `spdx identifier <https://spdx.org/licenses/>`_
     or a relative path to the license file.

Job definitions
---------------

``series`` (required)
     The name of the Ubuntu series used to run the job.

     Series names are :ref:`identifiers <identifiers>`.

``architectures`` (required)
     An architecture or list of architectures on which to run the job.  If a
     list, this multiplies the job into one copy per architecture.

     Architecture names are :ref:`identifiers <identifiers>`.

``packages`` (optional)
    Packages to install using ``apt`` as dependencies of this job.

``package-repositories`` (optional)
    Repositories which will be added to the already existing ones in
    `/etc/apt/sources.list`.
    Also see the :ref:`package-repositories` section below.

``snaps`` (optional)
    Snaps to install as dependencies of this job.
    Also see the :ref:`snap-properties` section below.

``environment`` (optional)
    A mapping of environment variable names to values, to be set while
    running the job. Takes precedence over environment variables set by
    plugins. Can be overridden via CLI.

``plugin`` (optional)
    A plugin which will be used for this job. See :doc:`../plugins`

``root`` (optional)
    If ``true``, run shell commands declared by ``run-before``,
    ``run``, and ``run-after`` as root;
    if ``false``, run them as a non-root user (``_lpci``).
    Default value: ``true``.

``run-before`` (optional)
    A string (possibly multi-line) containing shell commands to run for this
    job prior to the main ``run`` section.

``run`` (optional)
    A string (possibly multi-line) containing shell commands to run for this
    job.

``run-after`` (optional)
    A string (possibly multi-line) containing shell commands to run for this
    job after the main ``run`` section.

``output`` (optional)
    See the :ref:`output-properties` section below.

``input`` (optional)
    See the :ref:`input-properties` section below.

``matrix`` (optional)
    A list of mappings, each of which is a partial job definition.  The
    final list of concrete jobs to run for this job name is constructed by
    taking a copy of the job definition for each item in ``matrix``,
    removing the ``matrix`` key itself, and updating it with the contents of
    each item in turn.

.. note::

    Plugins can define :ref:`plugin_configuration_keys`.

.. _output-properties:

Output properties
-----------------

``paths`` (optional)
    A list of `Path.glob
    <https://docs.python.org/3/library/pathlib.html#pathlib.Path.glob>`_
    patterns; any files matching these patterns at the end of a successful
    build will be gathered by the build manager and attached to the build in
    Launchpad.  Paths may not escape the parent directory of the build tree.
    (The parent directory is allowed in order to make life easier for build
    systems such as ``dpkg-buildpackage`` that write output files to their
    parent directory.)

``distribute`` (optional)
    If ``artifactory``, then these artifacts may be distributed via
    Artifactory.

    Other valid values for ``distribute`` may be added in future.

``channels`` (optional)
    A list of initial channels to which these artifacts should be published
    (e.g. ``[edge]``).

``properties`` (optional)
    An arbitrary key/value mapping.  For Artifactory publication, these are
    attached as artifact properties; appropriate values depend on the
    package type.  Example properties include the human-readable version of
    the artifact.

``dynamic-properties`` (optional)
    A path (which may not escape the build tree), read using `python-dotenv
    <https://pypi.org/project/python-dotenv/>`_ and supplementing
    ``properties`` with the result.

``expires`` (optional)
    The requested minimum lifetime of the artifact in Launchpad.  Only
    relevant if ``distribute`` is not set or the artifact has not been
    successfully uploaded; an artifact that has been successfully uploaded
    is immediately eligible for garbage-collection from Launchpad, since it
    now exists elsewhere.

    This value is parsed using `pydantic's standard timedelta parsing
    <https://pydantic-docs.helpmanual.io/usage/types/#datetime-types>`_,
    restricted to non-negative timedeltas.

.. _input-properties:

Input properties
----------------

Input makes artifacts from previous pipeline stages available.  This only
works if those artifacts were saved using the ``--output-directory`` option
to ``lpci run``.

``lpci`` copies artifact data to the ``files`` subdirectory of the
designated target directory, and writes a ``properties`` file in the
designated target directory with JSON-encoded properties of the copied
artifacts.  (This mirrors the output file structure created by ``lpci run
--output-directory``.)

``job-name``
    The name of a previously-executed job whose artifacts should be made
    available.

``target-directory``
    A path, relative to the build tree of a project, identifying a directory
    to which the artifacts of the chosen job will be copied; the directory
    will be created if necessary.  Paths may not escape the build tree.

.. _snap-properties:

Snap properties
-----------------

``name``
    The name of the snap to install.

``channel`` (optional)
    `Channel <https://snapcraft.io/docs/channels>`_
    defining which release of a snap to install.
    Default value: ``latest/stable``.

``classic`` (optional)
    If True, allow installing snaps that use
    `classic confinement
    <https://snapcraft.io/docs/snap-confinement>`_.
    Default value: False.

.. note::

    To ensure backward compatibility default value for
    classic is True when defining snaps as a list of strings.
    (i.e. [chromium, firefox]).
    Please note that this format for defining snaps is deprecated.

.. _package-repositories:

Package-repositories properties
-------------------------------

The properties are inspired by the properties of `Snapcraft
<https://snapcraft.io/docs/package-repositories>`_.
Only a subset of them is currently implemented. More
properties can be implemented on demand.

A ``PPA`` or a ``deb`` repository can be added using the below properties.

Adding a PPA
^^^^^^^^^^^^

``type`` (required)
    Specifies the type of package-repository.
    Currently only ``apt`` is supported.

``formats`` (optional)
    Specifies the format of the package-repository.
    Supported values: ``deb`` and ``deb-src``. If unspecified,
    the format is assumed to be ``deb`` , i.e. ``[deb]``

``suites`` (optional)
    Specifies the suite of the package-repository.
    One or several of ``bionic``, ``focal``, ``jammy``. If unspecified,
    the suite is assumed to be the corresponding job's ``series`` value.

``ppa`` (required)
    Specifies the PPA to be used as the package repository in the short form,
    e.g. ``launchpad/ppa``, ``launchpad/debian/ppa``.

``trusted`` (optional)
    Set this to ``true`` to override APT's security checks, ie accept sources
    which do not pass authentication checks. ``false`` does the opposite.
    By default APT decides whether a source is considered trusted. This third
    option cannot be set explicitly.

Example:

.. code:: yaml

   package-repositories:
       - type: apt
         formats: [deb, deb-src]
         suites: [focal]
         ppa: launchpad/ubuntu/ppa
         trusted: false

Adding a deb repository
^^^^^^^^^^^^^^^^^^^^^^^

``type`` (required)
    Specifies the type of package-repository.
    Currently only ``apt`` is supported.

``formats`` (required)
    Specifies the format of the package-repository.
    Supported values: ``deb`` and ``deb-src``. If unspecified,
    the format is assumed to be ``deb``, i.e. ``[deb]``.

``suites`` (required)
    Specifies the suite of the package-repository.
    One or several of ``bionic``, ``focal``, ``jammy``. If unspecified,
    the suite is assumed to be the corresponding job's ``series`` value.

``components`` (required)
    Specifies the component of the package-repository,
    One or several of ``main``, ``restricted``, ``universe``, ``multiverse``.

``url`` (required)
    Specifies the URL of the package-repository,
    e.g. ``http://ppa.launchpad.net/snappy-dev/snapcraft-daily/ubuntu``.
    The URL is rendered using `Jinja2 <https://pypi.org/project/Jinja2/>`_.
    This can be used to supply authentication details via the *secrets*
    command line option.

``trusted`` (optional)
    Set this to ``true`` to override APT's security checks, ie accept sources
    which do not pass authentication checks. ``false`` does the opposite.
    By default APT decides whether a source is considered trusted. This third
    option cannot be set explicitly.

Example:

.. code:: yaml

   package-repositories:
       - type: apt
         formats: [deb, deb-src]
         components: [main]
         suites: [focal]
         url: https://canonical.example.org/ubuntu
         trusted: false

.. _license-properties:

License properties
------------------

Please note that either `spdx` or `path` is required.

``spdx`` (optional)
     A string representing a license,
     see `spdx identifier <https://spdx.org/licenses/>`_.

``path`` (optional)
    A string with the relative path to the license file.


NEWS.rst
===============
Version history
===============

0.2.3 (2023-07-20)
==================

- Upgrade ``PyYAML`` to 6.0.1 to fix https://github.com/yaml/pyyaml/issues/601,
  which was causing errors during the installation of ``PyYAML``.

0.2.2 (2023-07-14)
==================

- Fix default value for the `root` flag in documentation.

- Fix conda-build plugin to install and run outside
  of a conda environment.

0.2.1 (2023-06-06)
==================

- Rebuild the Snap package to include updated system packages with fixes
  for the following vulnerabilities.

  * https://ubuntu.com/security/notices/USN-6112-2/
  * https://ubuntu.com/security/notices/USN-6138-1/
  * https://ubuntu.com/security/notices/USN-6139-1/

0.2.0 (2023-05-24)
==================

- Add support for non-LTS devel release, which is currently mantic.

- Update Python dependencies, especially `Cryptography`, see CVE-2023-0286 and
  CVE-2023-23931.

- Fix Snap build issues for lpci on platforms which do not offer pre-built
  wheels for Cryptography.

0.1.2 (2023-05-02)
==================

- Rebuild the Snap package to include updated system packages
  https://ubuntu.com/security/notices/USN-6050-1/.

- Add ``dev0`` suffix documentation.

0.1.1 (2023-04-21)
==================
- Fix the yaml formatting in the snap deprecated format warning.

- Fix ``policy-rc.d`` issue preventing services
  from running into the container.

- Add a ``root`` flag.

0.1.0 (2023-04-18)
==================

- Add a ``release`` command.

- Rename project to ``lpci``.

- Add a ``--debug-shell`` flag.

- Add a ``recipe_folder`` parameter to specify
  the recipe itself, or the recipe search path
  using conda-build-plugin.

0.0.52 (2023-04-06)
===================
- Fix regression from adding support to snap keys
  passed as strings. For these keys confinement classic
  must be True by default to ensure backward compatibility.

0.0.51 (2023-04-05)
===================
- Add support for snap's channel and confinement level.
  See https://snapcraft.io/docs/channels and
  https://snapcraft.io/docs/snap-confinement.

0.0.50 (2023-03-20)
===================
- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5960-1.

0.0.49 (2023-03-10)
===================
- Fix regression from adding support for non-LTS releases. With the latest
  release of `craft-providers` we need to explicitly add a remote for images.

0.0.48 (2023-03-10)
===================

- Add support for non-LTS releases, that is currently lunar and kinetic.
  Please note that we use daily cloud images
  (https://cloud-images.ubuntu.com/daily/) for this purpose,
  so we cannot guarantee stability.

- Fix various ``mypy`` errors.

0.0.47 (2023-03-01)
===================

- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5821-3/.


0.0.46 (2023-02-28)
===================

- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5891-1/.

- Update dependencies.

- Fix deprecation warning in `lxd.launch` (changed API in `craft-providers`).

0.0.45 (2023-02-15)
===================

- Replace deprecated setuptools directive.
  See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-license-file.

- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5871-1/.

0.0.44 (2023-02-09)
===================

- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5849-1/.

0.0.43 (2023-01-24)
===================

- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5817-1.

0.0.42 (2023-01-20)
===================

- Add experimental ``--gpu-nvidia`` option.

0.0.41 (2023-01-18)
===================

- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5810-1/.

0.0.40 (2023-01-13)
===================

- Fix the leakage of package repositories from a job to the next.
- Add support for Python 3.11.
- Set sensible default values for some package repository fields.
- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5800-1.

0.0.39 (2023-01-06)
===================

- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5788-1.

0.0.38 (2023-01-05)
===================

- Allow specifying PPAs using the shortform notation,
  e.g. `ppa:launchpad/ubuntu/ppa`.

- Automatically import the signing keys for PPAs specified using
  the short-form notation.

0.0.37 (2022-12-09)
===================

- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5767-1.

0.0.36 (2022-12-08)
===================

- Sanitize the project name before cleaning.
- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5766-1/.

0.0.35 (2022-10-27)
===================

- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5702-1/.

0.0.34 (2022-10-20)
===================

- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5689-1.

0.0.33 (2022-10-19)
===================

- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5686-1.

0.0.32 (2022-10-14)
===================

- Rebuild the Snap package to include updated system packages.
  See https://ubuntu.com/security/notices/USN-5675-1.

0.0.31 (2022-09-12)
===================

- Move project directory from ``/root/lpcraft/project`` to
  ``/build/lpcraft/project``, making it more practical to drop privileges.

- Upgrade to craft-providers 1.4.2.

0.0.30 (2022-09-05)
===================


lpci/config.py
# Copyright 2021-2022 Canonical Ltd.  This software is licensed under the
# GNU General Public License version 3 (see the file LICENSE).

import re
from datetime import timedelta
from enum import Enum
from pathlib import Path
from typing import Any, Dict, Iterator, List, Optional, Tuple, Type, Union

import pydantic
from craft_cli import emit
from pydantic import AnyHttpUrl, StrictStr, root_validator, validator

from lpci.errors import ConfigurationError
from lpci.plugins import PLUGINS
from lpci.plugins.plugins import BaseConfig, BasePlugin
from lpci.utils import load_yaml

LAUNCHPAD_PPA_BASE_URL = "https://ppa.launchpadcontent.net"


class _Identifier(pydantic.ConstrainedStr):
    """A string with constrained syntax used as a short identifier.

    Compare `lp.app.validators.name` in Launchpad, though we also permit
    underscores here.
    """

    strict = True
    regex = re.compile(r"^[a-z0-9][a-z0-9\+\._\-]+$")


class ModelConfigDefaults(
    pydantic.BaseModel,
    extra=pydantic.Extra.forbid,
    alias_generator=lambda s: s.replace("_", "-"),
    underscore_attrs_are_private=True,
):
    """Define lpci's model defaults."""


class OutputDistributeEnum(Enum):
    """Valid values for `output.distribute`."""

    artifactory = "artifactory"


class Output(ModelConfigDefaults):
    """Job output properties."""

    paths: Optional[List[StrictStr]] = None
    distribute: Optional[OutputDistributeEnum] = None
    channels: Optional[List[StrictStr]] = None
    # instead of `Any` this should be something like `JSONSerializable`
    properties: Optional[Dict[StrictStr, Any]] = None
    dynamic_properties: Optional[Path] = None
    expires: Optional[timedelta] = None

    @pydantic.validator("expires")
    def validate_expires(cls, v: timedelta) -> timedelta:
        if v < timedelta(0):
            raise ValueError("non-negative duration expected")
        return v


class Input(ModelConfigDefaults):
    """Job input properties."""

    job_name: StrictStr
    target_directory: StrictStr


def _validate_plugin_config(
    plugin: Type[BasePlugin],
    values: Dict[StrictStr, Any],
    job_fields: List[str],
) -> Dict[StrictStr, Any]:
    plugin_config = {}
    for k in plugin.Config.schema()["properties"].keys():
        # configuration key belongs to the plugin
        if k in values and k not in job_fields:
            # TODO: should some error be raised if a plugin tries consuming
            # a job configuration key?
            plugin_config[k] = values.pop(k)
    values["plugin-config"] = plugin.Config.parse_obj(plugin_config)
    return values


class PackageType(str, Enum):
    """Specifies the type of the package repository.

    Currently only supports apt.
    """

    apt = "apt"


class PackageFormat(str, Enum):
    """Specifies the format of the package repository."""

    deb = "deb"
    deb_src = "deb-src"


class PackageComponent(str, Enum):
    """Specifies the component of the package repository."""

    main = "main"
    restricted = "restricted"
    universe = "universe"
    multiverse = "multiverse"


class PackageSuite(str, Enum):
    """Specifies the suite of the package repository.

    e.g. xenial, focal, ...
    """

    # XXX jugmac00 2023-03-10 the intention of this class was to verify that
    # only supported distroseries are present in the .launchpad.yaml file
    # but this does not work as intended, as you can specify arbitrary
    # strings which later possibly result in a KeyError
    bionic = "bionic"  # 18.04
    focal = "focal"  # 20.04
    jammy = "jammy"  # 22.04


class PPAShortFormURL(pydantic.ConstrainedStr):
    """A string with a constrained syntax to match a PPA short form URL."""

    strict = True
    # Support the two-segment form: OWNER/ARCHIVE (e.g. `launchpad/ppa`)
    # and the three-segment form: OWNER/DISTRIBUTION/ARCHIVE
    # (e.g. `launchpad/debian/ppa`).
    regex = re.compile(
        r"^[a-z0-9][a-z0-9\+\._\-]+(/[a-z0-9][a-z0-9\+\._\-]+){1,2}$"
    )


def get_ppa_url_parts(ppa_url: PPAShortFormURL) -> Tuple[str, str, str]:
    """Split and return the parts of a PPA short-form URL."""
    ppa_url_parts = ppa_url.split("/")
    if len(ppa_url_parts) == 2:
        owner, archive = ppa_url_parts
        distribution = "ubuntu"
    else:
        owner, distribution, archive = ppa_url_parts
    return owner, distribution, archive


class PackageRepository(ModelConfigDefaults):
    """A representation of a package repository.

    inspired by https://snapcraft.io/docs/package-repositories
    """

    type: PackageType  # e.g. `apt``
    ppa: Optional[PPAShortFormURL] = None  # e.g. `launchpad/ubuntu/ppa`
    formats: Optional[List[PackageFormat]] = None  # e.g. `[deb, deb-src]`
    components: Optional[List[PackageComponent]] = None  # e.g. `[main]`
    suites: Optional[List[PackageSuite]] = None  # e.g. `[bionic, focal]`
    url: Optional[AnyHttpUrl] = None
    trusted: Optional[bool] = False

    @root_validator(pre=True)
    def validate_multiple_fields(
        cls, values: Dict[str, Any]
    ) -> Dict[str, Any]:
        if "url" in values:
            if "ppa" in values:
                raise ValueError(
                    "Only one of the following keys can be specified:"
                    " 'url', 'ppa'."
                )
            if "components" not in values:
                raise ValueError(
                    "The 'components' key is required when the 'url' key"
                    " is specified."
                )
        else:
            if "ppa" not in values:
                raise ValueError(
                    "One of the following keys is required with an appropriate"
                    " value: 'url', 'ppa'."
                )
            if "components" in values:
                raise ValueError(
                    "The 'components' key is not allowed when the 'ppa' key is"
                    " specified. PPAs only support the 'main' component."
                )

        return values

    @validator("components", pre=True, always=True)
    def infer_components_if_ppa_is_set(
        cls, v: List[PackageComponent], values: Dict[str, Any]
    ) -> List[PackageComponent]:
        if v is None and values["ppa"]:
            return ["main"]
        return v

    @validator("url", pre=True, always=True)
    def infer_url_if_ppa_is_set(
        cls, v: AnyHttpUrl, values: Dict[str, Any]
    ) -> AnyHttpUrl:
        if v is None and values["ppa"]:
            owner, distribution, archive = get_ppa_url_parts(values["ppa"])
            v = "{}/{}/{}/{}".format(
                LAUNCHPAD_PPA_BASE_URL,
                owner,
                archive,
                distribution,
            )
        return v

    @validator("formats", pre=True, always=True)
    def set_formats_default_value(
        cls, v: List[PackageFormat]
    ) -> List[PackageFormat]:
        if not v:
            v = [PackageFormat.deb]
        return v

    @validator("trusted")
    def convert_trusted(cls, v: bool) -> str:
        # trusted is True or False, but we need `yes` or `no`
        return v and "yes" or "no"

    def sources_list_lines(self) -> Iterator[str]:
        """Yield repository lines as strings.

        e.g. 'deb https://canonical.example.org/artifactory/jammy-golang-backport focal main'
        """  # noqa: E501
        assert self.formats is not None
        assert self.suites is not None
        for format in self.formats:
            for suite in self.suites:
                assert self.components is not None
                if self.trusted:
                    yield f"{format.value} [trusted={self.trusted}] {self.url!s} {suite.value} {' '.join(self.components)}"  # noqa: E501
                else:
                    yield f"{format.value} {self.url!s} {suite.value} {' '.join(self.components)}"  # noqa: E501


class Snap(ModelConfigDefaults):
    """A snap definition."""

    name: StrictStr
    channel: Optional[StrictStr] = "latest/stable"
    classic: Optional[bool] = False

    @validator("channel")
    def prevent_channel_none(cls, v: StrictStr) -> Any:
        if v is None:
            raise ValueError(
                "You configured a Snap `channel`, "
                + "but you did not specify a value."
            )
        return v

    @validator("classic")
    def prevent_classic_none(cls, v: bool) -> Any:
        if v is None:
            raise ValueError(
                "You configured a Snap `classic`, "
                + "but you did not specify a value. "
                + "Valid values would either be `True` or `False`."
            )
        return v


class Job(ModelConfigDefaults):
    """A job definition."""

    # XXX jugmac00 2012-12-17: working with Job's attributes could be
    # simplified if they would not be optional, but rather return e.g.
    # an empty list

    series: _Identifier
    architectures: List[_Identifier]
    run_before: Optional[StrictStr]
    run: Optional[StrictStr]
    run_after: Optional[StrictStr]
    environment: Optional[Dict[str, Optional[str]]]
    output: Optional[Output]
    input: Optional[Input]
    snaps: Optional[List[Snap]]
    packages: Optional[List[StrictStr]]
    package_repositories: List[PackageRepository] = []
    plugin: Optional[StrictStr]
    plugin_config: Optional[BaseConfig]
    root: Optional[bool] = True

    @pydantic.validator("architectures", pre=True)
    def validate_architectures(
        cls, v: Union[_Identifier, List[_Identifier]]
    ) -> List[_Identifier]:
        if isinstance(v, str):
            v = [v]
        return v

    @pydantic.validator("root", pre=True)
    def validate_root(cls, v: Any) -> Any:
        if type(v) is not bool or v is None:
            raise ValueError(
                "You configured `root` parameter, "
                + "but you did not specify a valid value. "
                + "Valid values would either be `true` or `false`."
            )
        return v

    @pydantic.validator("snaps", pre=True)
    def validate_snaps(cls, v: List[Any]) -> Any:
        clean_values = []
        for value in v:
            # Backward compatibility, i.e. [chromium, firefox]
            if type(value) is str:
                emit.message(
                    f"Warning: You configured snap `{value}` but "
                    + "you used a deprecated format. "
                    + "\nPlease use "
                    + f"\n...\n- name: {value}\n"
                    + "  classic: true\n...\n"
                    + "instead.\n"
                    + "Please refer to the documentation for an "
                    + "overview of supported formats.",
                )
                clean_values.append({"name": value, "classic": True})
            elif type(value) is dict:
                if "name" not in value or value["name"] is None:
                    raise ValueError(
                        "You configured a Snap "
                        + "but you did not specify a name."
                    )
                if "classic" in value and value["classic"] is not None:
                    if type(value["classic"]) is not bool:
                        raise ValueError(
                            "You configured a Snap `classic`, "
                            + "but you did not specify a valid value. "
                            + "Valid values would either be `true` or `false`."
                        )
                clean_values.append(value)
            else:
                raise ValueError(
                    "You configured a Snap, "
                    + "but you used an unknown format. "
                    + "Please refer to the documentation for an "
                    + "overview of supported formats."
                )
        return clean_values

    @pydantic.validator("package_repositories")
    def validate_package_repositories(
        cls, v: List[PackageRepository], values: Dict[StrictStr, Any]
    ) -> List[PackageRepository]:
        package_repositories = None
        for index, package_repository in enumerate(v):
            if not package_repository.suites:
                if not package_repositories:
                    package_repositories = v.copy()
                package_repositories[index].suites = [
                    PackageSuite[values["series"]]
                ]
        return package_repositories or v

    @pydantic.root_validator(pre=True)
    def move_plugin_config_settings(
        cls, values: Dict[StrictStr, Any]
    ) -> Dict[StrictStr, Any]:
        """Delegate plugin settings to the plugin."""
        if "plugin" in values:
            base_values = values.copy()
            if values["plugin"] not in PLUGINS:
                raise ConfigurationError("Unknown plugin")
            plugin = PLUGINS[values["plugin"]]
            return _validate_plugin_config(
                plugin=plugin,
                values=base_values,
                job_fields=list(cls.__fields__.keys()),
            )
        return values


def _expand_job_values(
    values: Dict[StrictStr, Any]
) -> List[Dict[StrictStr, Any]]:
    expanded_values = []
    if "matrix" in values:
        base_values = values.copy()
        del base_values["matrix"]
        for variant in values["matrix"]:
            variant_values = base_values.copy()
            variant_values.update(variant)
            expanded_values.append(variant_values)
    else:
        expanded_values.append(values)
    return expanded_values


class License(ModelConfigDefaults):
    """A representation of a license."""

    # We do not need to check that at least one value is set, as currently
    # there are only these two values, no others. That means not setting any of
    # them will not result in the creation of a `License` object.
    # Once we have more fields, we need to add e.g. a root validator, see
    # https://stackoverflow.com/questions/58958970

    # XXX jugmac00 2022-08-03: add validator for spdx identifier
    # XXX jugmac00 2022-08-04: add validator for path

    spdx: Optional[StrictStr] = None
    path: Optional[StrictStr] = None

    @validator("path", always=True)
    def disallow_setting_both_sources(
        cls, path: str, values: Dict[str, str]
    ) -> str:
        if values.get("spdx") and path:
            raise ValueError(
                "You cannot set `spdx` and `path` at the same time."
            )
        return path


class Config(ModelConfigDefaults):
    """A .launchpad.yaml configuration file."""

    pipeline: List[List[_Identifier]]
    jobs: Dict[StrictStr, List[Job]]
    license: Optional[License]

    @pydantic.validator("pipeline", pre=True)
    def validate_pipeline(
        cls, v: List[Union[_Identifier, List[_Identifier]]]
    ) -> List[List[_Identifier]]:
        return [[stage] if isinstance(stage, str) else stage for stage in v]

    # XXX cjwatson 2021-11-17: This expansion strategy works, but it may
    # produce suboptimal error messages, and doesn't have a good way to do
    # things like limiting the keys that can be set in a matrix.
    @pydantic.root_validator(pre=True)
    def expand_matrix(
        cls, values: Dict[StrictStr, Any]
    ) -> Dict[StrictStr, Any]:
        expanded_values = values.copy()
        expanded_values["jobs"] = {
            job_name: _expand_job_values(job_values)
            for job_name, job_values in values["jobs"].items()
        }
        return expanded_values

    @classmethod
    def load(cls, path: Path) -> "Config":
        """Load config from the indicated file name."""
        # XXX lgp171188 2022-03-31: it is not a good idea to evaluate
        # `Path.cwd()` in multiple places to determine the project directory.
        # This could be made available as an attribute on the `Config` class
        # instead.
        project_dir = Path.cwd()
        resolved_path = path.resolve()
        try:
            # XXX lgp171188 2022-04-04: There is a new method,
            # Path.is_relative_to() in Python 3.9+, which does
            # exactly what we need. Once we drop support
            # for Python 3.8, we should switch to that instead.
            resolved_path.relative_to(project_dir)
        except ValueError:
            raise ConfigurationError(
                f"'{resolved_path}' is not in the subpath of '{project_dir}'."
            )
        content = load_yaml(path)
        return cls.parse_obj(content)
